### PR TITLE
Implement stencil fns for WebGLRenderingContext

### LINKED ISSUE #10659: Implement WebGLRenderingContext::stencil*
Part of #10209.

### PR DESCRIPTION
**Pull request by KiChjang**
Depends on servo/webrender#261.

Closes #10659.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10670)

### CODE REVIEW

File: components/script/dom/webglrenderingcontext.rs [hunk 1]
```diff
@@ -1084,6 +1084,48 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.3
+    fn StencilOp(&self, fail: u32, zfail: u32, zpass: u32) {
+        self.ipc_renderer
+            .send(CanvasMsg::WebGL(WebGLCommand::StencilOp(fail, zfail, zpass)))
+            .unwrap()
+    }
+
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.3
+    fn StencilOpSeparate(&self, face: u32, fail: u32, zfail: u32, zpass: u32) {
+        self.ipc_renderer
```
Comment (KiChjang) [inline components/script/dom/webglrenderingcontext.rs]: I'm not sure if these methods need to validate their parameters as described [here](https://www.khronos.org/registry/webgl/specs/latest/1.0/#STENCIL_SEPARATE_LIMIT).

File: components/script/dom/webglrenderingcontext.rs [hunk 2]
```diff
@@ -1084,6 +1084,48 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
             .unwrap()
     }
 
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.3
+    fn StencilFunc(&self, func: u32, ref_: i32, mask: u32) {
```
Comment (ConnorGBrewster) [inline components/script/dom/webglrenderingcontext.rs]: I think these should validate the `func` parameter as described [here](https://www.khronos.org/opengles/sdk/docs/man/xhtml/glStencilFunc.xml)

File: components/script/dom/webglrenderingcontext.rs (hunk 1 above)
Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Yes, all parameters should be validated first taking in account the [GLES2.0 spec][gles], and later taking in account any details that the [WebGL spec][webgl] could specify.

[gles]: https://www.khronos.org/opengles/sdk/docs/man/
[webgl]: https://www.khronos.org/registry/webgl/specs/latest/1.0/

File: components/script/dom/webglrenderingcontext.rs (hunk 2 above)
Comment (KiChjang) [inline components/script/dom/webglrenderingcontext.rs]: In which case, should these functions have a [Throws] WebIDL attribute added to them?

Comment (emilio) [inline components/script/dom/webglrenderingcontext.rs]: Nope! WebGL errors are tracked via `gl.getError()`, not via return values or exceptions.

Comment (ConnorGBrewster) [inline components/script/dom/webglrenderingcontext.rs]: I do not think so, none of the other functions that can result in a WebGL error seem to have a  [Throws] WebIDL attribute.

**Comment by KiChjang** [comment]
Added checks for enum values with the stencil functions.

File: components/script/dom/webglrenderingcontext.rs [hunk 3]
```diff
@@ -1084,6 +1084,106 @@ impl WebGLRenderingContextMethods for WebGLRenderingContext {
...
+    }
+
+    // https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.3
+    fn StencilOpSeparate(&self, face: u32, fail: u32, zfail: u32, zpass: u32) {
+        match face {
+            constants::FRONT | constants::BACK | constants::FRONT_AND_BACK => (),
+            _ => return self.webgl_error(InvalidEnum),
+        }
+
+        match fail {
```
Comment (dzbarsky) [inline components/script/dom/webglrenderingcontext.rs]: it might be nice if these were factored out

**Comment by KiChjang** [comment]
Created a new function for checking stencil op parameters.

**Comment by KiChjang** [comment]
Is now blocked on #10678.

**Comment by KiChjang** [comment]
Should be ok to review now.

**Comment by emilio** [comment]
Looks good to me, I usually prefer the early return approach instead of if else, but feel free to r=me if you think it's not worth the change.

@bors-servo: try

---

Reviewed 1 of 5 files at r1, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10670#-:-KFjYAn53WkTIWlpOudf:913352929)*

**Comment by jdm** [comment]
@bors-servo: retry
* #10473

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:
  │ FAIL [expected PASS] WebGL test #24: contextA.getUniform(programS, locationSx) should be 3. Threw exception TypeError: contextA.getUniform is not a function
  │   → assert_true: contextA.getUniform(programS, locationSx) should be 3. Threw exception TypeError: contextA.getUniform is not a function expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:1
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:229:9
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:61:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:
  │ FAIL [expected PASS] WebGL test #23: getError expected: NO_ERROR. Was INVALID_OPERATION : after evaluating: contextA.uniform1f(locationArray0, 123)
  │   → assert_true: getError expected: NO_ERROR. Was INVALID_OPERATION : after evaluating: contextA.uniform1f(locationArray0, 123) expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:1
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ glErrorShouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1379:5
  │ shouldGenerateGLError@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1355:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:60:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:
  │ FAIL [expected PASS] WebGL test #25: contextA.getUniform(programS, locationArray0) should be 123. Threw exception TypeError: contextA.getUniform is not a function
  │   → assert_true: contextA.getUniform(programS, locationArray0) should be 123. Threw exception TypeError: contextA.getUniform is not a function expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:1
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ shouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:229:9
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:62:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:
  │ FAIL [expected PASS] WebGL test #20: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.uniform1i(locationSx, 3)
  │   → assert_true: getError expected: INVALID_OPERATION. Was NO_ERROR : after evaluating: contextA.uniform1i(locationSx, 3) expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:1
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ glErrorShouldBe@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1379:5
  │ shouldGenerateGLError@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1355:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:51:1

  ▶ Unexpected subtest result in /webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:
  │ FAIL [expected PASS] WebGL test #21: contextA.getUniform(programS, locationSx) threw exception TypeError: contextA.getUniform is not a function
  │   → assert_true: contextA.getUniform(programS, locationSx) threw exception TypeError: contextA.getUniform is not a function expected true got false
  │ 
  │ reportTestResultsToHarness/<@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:87:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1397:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ reportTestResultsToHarness@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:86:1
  │ testFailed@http://web-platform.test:8000/webgl/conformance-1.0.3/resources/js-test-pre.js:152:5
  │ shouldGenerateGLError@http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/resources/webgl-test-utils.js:1353:5
  └ @http://web-platform.test:8000/webgl/conformance-1.0.3/conformance/uniforms/uniform-location.html:52:1
```

**Comment by KiChjang** [comment]
@bors-servo retry
* #10790

**Comment by KiChjang** [comment]
@bors-servo retry
* #10753

**Comment by jdm** [comment]
It seems likely that #10790 is not actually intermittent and is a permanent failure exposed by the current set of changes.

**Comment by jdm** [comment]
Yeah, it's consistently failed on OS X every time there hasn't been a linux failure first.

**Comment by emilio** [comment]
That test should be fixed by #10443.

Look at [this comment](https://github.com/servo/servo/pull/10443#issuecomment-212586661) though, to see why new correct implementations can lead to new test failures.

**Comment by emilio** [comment]
I'd rather have waited for #10806 to land... But well, it landed so... :P

**Pull request merged**